src/Npgsql.SourceGenerators/DatabaseInfoSourceGenerator.cs

```diff
@@ -0,0 +1,643 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Dynamic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using Scriban;
+using Scriban.Runtime;
+
+namespace Npgsql.SourceGenerators;
+
+[Generator]
+sealed class DatabaseInfoSourceGenerator : ISourceGenerator
+{
+    // Warning: Do not change to target typing
+    // See: https://github.com/dotnet/roslyn-analyzers/issues/5890#issuecomment-1046043775
+    // ReSharper disable once ArrangeObjectCreationWhenTypeEvident
+    static readonly DiagnosticDescriptor DatFileParserError = new DiagnosticDescriptor(
+        id: "NPGSQLDBINFOGEN001",
+        title: "Failed to parse .dat file",
+        messageFormat: "Syntax error in .dat file",
+        category: "DatabaseInfoSourceGenerator",
+        DiagnosticSeverity.Error,
+        isEnabledByDefault: true);
+
+    public void Initialize(GeneratorInitializationContext context) { }
+
+    public void Execute(GeneratorExecutionContext context)
+    {
+
+        List<DatFileObject>? pgType = null;
+        List<DatFileObject>? pgRange = null;
+        var noDatFile = true;
+        foreach (var file in context.AdditionalFiles)
+        {
+            if (Path.GetExtension(file.Path) != ".dat")
+                continue;
+
+            noDatFile = false;
+            var fileName = Path.GetFileName(file.Path);
+            try
+            {
+                switch (fileName)
+                {
+                case "pg_type.dat":
+                    pgType = ParseDatFile(file.GetText());
+                    break;
+                case "pg_range.dat":
+                    pgRange = ParseDatFile(file.GetText());
+                    break;
+                default:
+                    throw new InvalidOperationException($"This source generator is not designed to make use of the information in '{fileName}'. " +
```

Comment:
Technically, we should maybe not fail if an unknown .dat file is present (but just pick up those we know), but that's very very minor.